### DISCUSSION TITLE
Grid Columns Width Not Equal On All Rows

### DISCUSSION THREAD

**Discussion by mysteryx93** [AvaloniaUI/Avalonia#9116] [Q&A] [answered]
I'm using this code instead of a DataGrid to show read-only data. Much lighter.

The problem is that if the text in Column 0 is larger than the column, it moves the text in Column 1 to the right, so the columns are all uneven. How can I make the columns align on all rows?

```c#
<ScrollViewer Margin="0,20,0,0">
    <ItemsRepeater Items="{Binding Encoder.ProcessingFiles}">
        <ItemsRepeater.ItemTemplate>
            <DataTemplate DataType="{x:Type bass:ProcessingItem}">
                <Grid ColumnDefinitions="*,120,8">
                    <TextBlock Grid.Column="0" Text="{Binding Text}" Margin="0,0,8,0" />
                    <ProgressBar Grid.Column="1" Value="{Binding ProgressPercent, Mode=OneWay}"
                                 Maximum="1" Height="24" Background="Transparent" MinWidth="0"
                                 Foreground="#0078d7" />
                    <TextBlock Grid.Column="1" Text="{Binding Status, Mode=OneWay}"
                               VerticalAlignment="Center" Margin="12,0,0,0" />
                    <Grid Column="2" MinWidth="8" />
                </Grid>
            </DataTemplate>
        </ItemsRepeater.ItemTemplate>
    </ItemsRepeater>
</ScrollViewer>
```

![Screenshot_432hz Batch Converter_1](https://user-images.githubusercontent.com/8528033/194361101-4d3211a7-a3fe-4d59-8992-67cb538baf1c.png)

**Comment by maxkatz6** [comment]
As it's expected for the grid, because these grids are not synchronized with each other.
You can use SharedSizeGroup to share same size for the grid,

Other way can be correcting the alignment of the grid. It looks like it is left aligned right now, but you can set grid alignment to stretch. Maybe add alignment to repeater itself too.

**Reply by maxkatz6** [reply]
Specifically for your task grid alignment probably is the best straightforward solution

**Reply by mysteryx93** [reply]
Adding `HorizontalAlignment="Stretch"` doesn't change anything. `SharedSizeGroup` is designed to sync 2 separate grids; would be overkill for within the same grid.

**Reply by workgroupengineering** [reply]
each item of ItemsRepeater  is grid, you can see in [DevTools](https://docs.avaloniaui.net/docs/getting-started/developer-tools).

**Reply by mysteryx93** [reply]
AAAAHHH!! So they ARE in fact separate grids. You are right.

Is that the best solution? Or is there a way to generate it as a single grid?

**Reply by timunie** [reply]
You can share gridsize over multiple grids if you want to. 

See this for a WPF sample. In Avalonia it works quite similar: https://bengribaudo.com/blog/2011/05/21/1228/wpf-separate-grids-auto-sized-columnsrows-with-synchronized-dimensions

**Comment by mysteryx93** [comment, marked as answer]
The grid still doesn't display right...

With SharedSizeGroup on the first column, that column is too large. Without it, the column is too small...

```xaml
<ItemsRepeater Items="{Binding Encoder.ProcessingFiles}" Grid.IsSharedSizeScope="True">
    <ItemsRepeater.ItemTemplate>
        <DataTemplate DataType="{x:Type bass:ProcessingItem}">
            <Grid>
                <Grid.ColumnDefinitions>
                    <ColumnDefinition SharedSizeGroup="ProcessingCol1" Width="*" />
                    <ColumnDefinition SharedSizeGroup="ProcessingCol2" Width="120" />
                    <ColumnDefinition SharedSizeGroup="ProcessingCol3" Width="8" />
                </Grid.ColumnDefinitions>
                <TextBlock Grid.Column="0" Text="{Binding Text}" Margin="0,0,8,0" />
                <ProgressBar Grid.Column="1" Value="{Binding ProgressPercent, Mode=OneWay}"
                             Maximum="1" Height="24" Background="Transparent" MinWidth="0"
                             Foreground="#0078d7" />
                <TextBlock Grid.Column="1" Text="{Binding Status, Mode=OneWay}"
                           VerticalAlignment="Center" Margin="12,0,0,0" />
                <Grid Column="2" MinWidth="8" />
            </Grid>
        </DataTemplate>
    </ItemsRepeater.ItemTemplate>
</ItemsRepeater>
```

Perhaps another solution would be to ensure that Column 1 & 2 always maintain their fixed size. Not sure why they don't, even if I set MinWidth. It seems to think that displaying the text is more important.

**Reply by timunie** [reply]
`*` means take up the entire remaining space. So as the other sizes are fixed as well, this all makes no sence except if the grid is not stretched horizontally. You can use DevTools `[F12]` to inspect the styling of each node.

**Reply by mysteryx93** [reply]
Setting `<Grid HorizontalAlignment="Stretch">` doesn't help either. I believe it's the default anyway.

**Reply by timunie** [reply]
Without a simple sample I can't help you further. Sorry.
